src/backend/distributed/operations/citus_tools.c

```diff
@@ -93,6 +101,122 @@ CheckConnectionToNode(char *nodeName, uint32 nodePort)
 }
 
 
+/*
+ * citus_check_cluster_node_health UDF performs connectivity checks from all the nodes to
+ * all the nodes, and report success status
+ */
+Datum
+citus_check_cluster_node_health(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = SetupTuplestore(fcinfo, &tupleDescriptor);
+
+	StoreAllConnectivityChecks(tupleStore, tupleDescriptor);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetConnectivityCheckCommand returns the command to check connections to a node
+ */
+static char *
+GetConnectivityCheckCommand(const char *nodeName, const uint32 nodePort)
+{
+	StringInfo connectivityCheckCommand = makeStringInfo();
+	appendStringInfo(connectivityCheckCommand,
+					 "SELECT citus_check_connection_to_node('%s', %d)",
+					 nodeName, nodePort);
+
+	return connectivityCheckCommand->data;
+}
+
+
+/*
+ * StoreAllConnectivityChecks performs connectivity checks from all the nodes to all the
+ * nodes, and report success status.
+ *
+ * Algorithm is:
+ * for sourceNode in activePrimaryWorkerList:
+ *   c = connectToNode(sourceNode)
+ *   for targetNode in activePrimaryWorkerList:
+ *     status = c.execute("SELECT citus_check_connection_to_node(targetNode.name, targetNode.port")
+ *     emit sourceNode.name, sourceNode.port, targetNode.name, targetNode.port, status
+ */
+static void
+StoreAllConnectivityChecks(Tuplestorestate *tupleStore, TupleDesc tupleDescriptor)
+{
+	Datum values[CONNECTIVITY_CHECK_COLUMNS];
+	bool isNulls[CONNECTIVITY_CHECK_COLUMNS];
+
+	List *workerNodeList = ActivePrimaryNodeList(ShareLock);
```

Comment:
What about read-replicas? It seems useful that this UDF works there as well.

I think we can probably use `ActiveReadableNodeList` instead of `ActivePrimaryNodeList`. But, we should also add one more test to somewhere like `multi_follower_select_statements.sql`

Reply:
Do you know how to setup Citus-ha with read-replicas? Once you change this, it'd be great to test. Let me know if you need any help testing this

Reply:
I ended up reading directly from the catalogs as our cache does not include workers from other clusters.

AFAIK we have some users that set the same `nodecluster` values of a set of read replicas in a cluster. Reading from catalog seemed to be the best way to cover connectivity checks to those nodes as well.

Reply:
> AFAIK we have some users that set the same nodecluster values of a set of read replicas in a cluste

Short answer: We should switch back to `ActiveReadableNodeList ` :) 

Long answer:

I really don't think this is a good idea & even exists. 

Even if you have multiple read-replicas, you can only query one of the clusters at a time. And, typically you cannot change this because it is `PGC_SU_BACKEND`. So, if you want to query the other replica, you should do:
```
ALTER SYSTEM SET citus.cluster_name TO ...
SELECT pg_reload_conf();
....
-- reconnect via psql
SELECT query....
```

Also, there is no mechanism in the code that can handle multiple read-replicas at a time. `ReadNodeList(true)` is only used once in the code, when metadata is syned. Also, `ClusterName` expects a single cluster name anyway.

Also, see how the planner/executor assigns task placements. It is always relied on as `ActiveReadableNodeList `. And, our aim with this UDF is to signal that query executions can be done fine when returns true.